crates/oxc_linter/src/rules/for_direction.rs

```diff
@@ -0,0 +1,156 @@
+            } else {
+                UNKNOWN
+            }
+        }
+        // can't determine other kinds of updates
+        _ => UNKNOWN,
+    }
+}
+
+fn get_assignment_direction(assign: &AssignmentExpression) -> UpdateDirection {
+    let operator = &assign.operator;
+    let right = &assign.right;
+    let positive = match right {
+        Expression::NumberLiteral(r) => r.value.is_sign_positive(),
+        Expression::UnaryExpression(right) => right.operator != UnaryOperator::UnaryNegation,
+        _ => return UNKNOWN,
+    };
+
+    let mut direction = match operator {
+        AssignmentOperator::Addition => FORWARD,
+        AssignmentOperator::Subtraction => BACKWARD,
+        _ => return UNKNOWN,
+    };
+
+    if !positive {
+        direction = -direction;
+    }
+    direction
+}
+
+impl Rule for ForDirection {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::ForStatement(for_loop) = node.get().kind() {
+            if let Some(Expression::BinaryExpression(test)) = &for_loop.test {
+                if let Expression::Identifier(counter) = &test.left {
+                    let test_operator = &test.operator;
+                    let wrong_direction = match test_operator {
+                        BinaryOperator::LessEqualThan | BinaryOperator::LessThan => BACKWARD,
+                        BinaryOperator::GreaterEqualThan | BinaryOperator::GreaterThan => FORWARD,
+                        _ => return,
+                    };
+                    if let Some(update) = &for_loop.update {
+                        let update_direction = get_update_direction(update, counter);
+                        if update_direction == wrong_direction {
+                            ctx.diagnostic(ForDirectionDiagnostic(for_loop.span));
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[test]
+fn test() {
+    use crate::tester::Tester;
+
+    let pass = vec![
+        ("for (var a = 0; a <= 10; a += 1) {}", None),
+        ("for (var a = 0; a <= 10; a -= -1) {}", None),
+        ("for (var a = 0; a <= 10; a++) {}", None),
+        ("for (var a = 0; a > 10; a -= 1) {}", None),
+        ("for (var a = 0; a > 10; a += -1) {}", None),
+        ("for (var a = 0; a > 10; a--) {}", None),
+        ("while (true) {}", None),
+        ("for (; a < 10; b--) {}", None),
+        ("for (; a > 10; b += 1) {}", None),
+        ("for (; lessThen(a, 1); a++) {}", None),
+        ("for (; a < 10; a -= foo()) {}", None),
+        ("for (; a < 10; a -= c) {}", None),
+    ];
+
+    let fail = vec![
+        ("for (var a = 0; a < 10; a--) {}", None),
+        ("for (var a = 0; a < 10; a += -1) {}", None),
+        ("for (var a = 0; a < 10; a -= 1) {}", None),
+        ("for (; a >= 0; a++) {}", None),
+        ("for (; a >= 0; a -= -1) {}", None),
+        ("for (; a >= 0; a += 1) {}", None),
+        ("for (; a < 10; a += -foo()) {}", None),
+    ];
```

Comment:
I think we should copy eslint's tests, so we don't get lost.

Reply:
Done.